### PR TITLE
Split type erasure logic from erased_serde logic in ErasedDataProvider

### PR DESCRIPTION
**Pull request by sffc**
As discussed in #563, the previous architecture where ErasedDataProvider was a "grab bag" that included both the type erasure logic (converting data structs to Any) and the trait object friendly Serde provider (via erased_serde) prevented ErasedDataProvider from using zero-copy deserialization.

To remedy this, I split ErasedDataProvider into two traits (better names welcome):

1. SerdeDataProvider: contains only the one Serde method
2. ErasedDataProvider: contains all other methods

I changed the new C API to use SerdeDataProvider instead of ErasedDataProvider.

I also took the opportunity to remove the data receiver from ErasedDataProvider, since the receiver design is really only needed when Serde is being used.

### CODE REVIEW

**Comment by sffc** [comment]
I pushed some more commits where I further split up ErasedDataStruct to remove the serde dependency from it, and put serde in a different struct called SerdeSeDataStruct ("SerdeSe" to emphasize that the struct supports serialization only).  Doing this also enables the new struct to have a non-static lifetime, which should help unblock #466.

**Comment by sffc** [comment]
Converting to draft because I'm still doing some more refactoring.

**Comment by sffc** [comment]
Okay, I can keep refactoring for ever, but I'm going to stop now. I've already filed a follow-up issue for another piece I want to refactor next quarter. :laughing: 

I've successfully decoupled Serde from ErasedDataProvider, meaning that we can support zero-copy deserialization in FFI, and we can also support serialization of providers that are non-static, like PPUCD.

**Comment by sffc** [comment]
I know it's a longish pull request, but:

1. There are more lines deleted than lines added
2. I am confident in the code
3. Many of the lines are new documentation

File: components/provider/src/serde.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+use crate::prelude::*;
+use std::borrow::Cow;
+use std::fmt::Debug;
+
+/// An object that receives data from a Serde Deserializer. Implemented by [DataPayload].
+///
+/// Lifetimes:
+///
+/// - `'de` = deserializer lifetime; can usually be `'_`
+pub trait SerdeDeDataReceiver<'de> {
```
Comment (Manishearth) [inline components/provider/src/serde.rs:32]: nit: can it just be SerdeDataReceiver? Not a strong opinion.

Comment (Manishearth) [reply components/provider/src/serde.rs:32]: I guess it's nice to be explicit

File: components/provider/src/serde.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    /// Consumes a Serde Deserializer into this SerdeDeDataReceiver as owned data.
+    ///
+    /// This method results in an owned payload, but the payload could have non-static references
```
Comment (Manishearth) [inline components/provider/src/serde.rs:35]: this is a clever way of handling this particular lifetime conundrum

**Review by Manishearth — APPROVED** [review]
Takes a while to review but the overall design looks good!

File: components/provider/src/export.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    /// Save a `payload` corresponding to the given data request (resource path).
+    fn put_payload(
+        &mut self,
+        req: &DataRequest,
+        payload: &T,
+    ) -> Result<(), Box<dyn std::error::Error>>;
+
+    /// Whether to load and dump data for the given entry. This function enables the
+    /// `DataExporter` to filter out certain data entries.
+    fn include_resource_options(&self, resc_options: &ResourceOptions) -> bool;
```
[the remark this answers is not in the document]
Comment (sffc) [reply components/provider/src/export.rs:29]: `should_include_...` is more appropriate than `will_include_...`, but neither is very good, so I went with the shorter name.

**Pull request merged**